### PR TITLE
Retain base languages for compiled data

### PR DESCRIPTION
**Pull request by robertbastian**
#58

### CODE REVIEW

File: provider/data/collator/data/collation_diacritics_v1_marker.rs.data [hunk 1]
```diff
@@ -4,7 +4,7 @@
 /// `icu`'s `_unstable` constructors.
 ///
 /// Using this implementation will embed the following data in the binary's data segment:
-/// * 41B for the lookup data structure (2 data identifiers)
+/// * 606B for the lookup data structure (151 data identifiers)
```
Comment (robertbastian) [inline provider/data/collator/data/collation_diacritics_v1_marker.rs.data:7]: not great

File: provider/data/datetime/data/buddhist_year_names_v1_marker.rs.data [hunk 2]
```diff
@@ -4,7 +4,7 @@
 /// `icu`'s `_unstable` constructors.
 ///
 /// Using this implementation will embed the following data in the binary's data segment:
-/// * 650B for the lookup data structure (117 data identifiers)
```
Comment (robertbastian) [inline provider/data/datetime/data/buddhist_year_names_v1_marker.rs.data:7]: not great

File: provider/data/plurals/data/plural_ranges_v1_marker.rs.data [hunk 3]
```diff
@@ -4,7 +4,7 @@
 /// `icu`'s `_unstable` constructors.
 ///
 /// Using this implementation will embed the following data in the binary's data segment:
-/// * 130B for the lookup data structure (25 data identifiers)
+/// * 631B for the lookup data structure (156 data identifiers)
```
Comment (robertbastian) [inline provider/data/plurals/data/plural_ranges_v1_marker.rs.data:7]: not great

**Review by robertbastian — COMMENTED** [review]
Overall it looks like it's roughly +50% for the lookup trie, in some cases it's much more, in few cases less. Highlighted some cases where it by far exceeds the previous size, or the size of the actual data.

**Comment by jedel1043** [comment]
> Overall it looks like it's roughly +50% for the lookup trie, in some cases it's much more, in few cases less. Highlighted some cases where it by far exceeds the previous size, or the size of the actual data.

Now that `DryDataProvider` is a thing, maybe we could move the duplicated locales to it, then always do full deduplication while fetching the full data? That could enable DCE for people who don't use `can_load`.

**Comment by robertbastian** [comment]
You mean having separate lookup tries for `dry_load` and `load`? That will mean duplication for people who use `dry_load`. It also means there will be different behaviour between `load` and `dry_load`

**Comment by jedel1043** [comment]
Yeah, but I also was thinking that the Rust compiler cannot deduplicate between `DataProvider` impls because they don't share the same data, but maybe the compiler could deduplicate between `DryDataProvider` impls since the metadata is basically the same between each other.

About the differing behaviour, I think it's fine. The trait just guarantees that if `DataProvider` doesn't fail, then `DryDataProvider` must not fail for the same request, but I think it wouldn't be out of place to not guarantee that the returned locale will be the same between them.

**Comment by sffc** [comment]
We had a hypothesis at some point that retain-base-languages might be more efficient since it saves a hop during locale fallbacking. Do we have figures for how much it saves on average? Maybe based on one of our examples?

**Comment by sffc** [comment]
I think @jedel1043's idea is worth investigating. Maybe a hybrid approach where there are 2 tries, one that has maximal deduplication lookup and the other that holds the diff between retain-base-languages and maximal.

**Comment by sffc** [comment]
Or actually, all of the DryDataProvider impls could just use the same trie, which always contains _only_ the base langauges, and all of the impls generated from the same datagen invocation can use the exact same trie.

**Comment by sffc** [comment]
Some discussion:

- @robertbastian - DryDataProvider can be achieved without retain-base-languages, this PR is basically a lookup size/speed tradeoff
- @sffc - I would generally trade a 10% size increase for a 25% speed increase, or a 1-2% size increase for a 5% speed increase. I would be happy if we made "load formatter and use it once" call sites faster, and data loading is still a nontrivial portion of the time spent in those call sites.
- @robertbastian - % of what? It really depends on the usage pattern, how many fallback vs non-fallback locales, which ICU4X APIs, etc
- @Manishearth - I don't think compiled data loading is 25% of a typical ICU4X usage, so Amdahl's law stops us from getting speedups that big. It may be for usage patterns where people never retain formatter objects ("load formatter and use it once"), but I'd like to see benchmarks, and people in those situations have other optimization tradeoffs available to them (cache the objects, which may give them a small size increase trading off for fast lookups)
- @sffc - We shouldn't add these impls unless we have base languages, because they are useless without base languages.
- @robertbastian - I don't want to conditionally add them based on deduplication strategy.
- @sffc - Then let's export the macros without calling the macros on `provider::Baked`.

**Comment by sffc** [comment]
This PR doesn't add DryDataProvider impls yet, right? It's just changing the deduplication mode?

**Review by sffc — CHANGES_REQUESTED** [review]
This PR hasn't changed since the last discussion. What is your expectation of my behavior? I still have open questions.

- What is the impact on metrics? This increases data size, sometimes quite substantially.
- We have paths toward supporting DryDataProvider that don't require this PR and require much less data.
- I'm not convinced that we want retain_base_languages behavior on the default Baked constructors. It is a niche use case. There was a hypothesis that it improves performance, but I haven't seen data to back that up.
- Also note that the DryDataProvider impls can't be used unless the client goes and looks at `icu::component::provider::Baked`, which is unstable and not generally something we wanted clients to interact with.

**Comment by sffc** [comment]
Example of a solution that might be more efficient (certainly a lot smaller code/data size):

- Add a single trie that contains all base languages
- DataProvider impls check that trie when they are processing a base language and didn't find an entry in their own trie; if the requested language/script is in the base language trie, then they return the `und` entry for that request

**Pull request closed without merging**